analyses/molecular-subtyping-EPN/03-subgrouping_samples.py

```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env python
+# Author Teja Koganti (D3B)
+
+# This script takes in the final EPN table with all the columns (in results folder  results/EPN_all.tsv), 
+#	sets some rules and thresholds so samples can be  subgrouped
+#	and adds the sub group name in the last "subgroup" column
+
+## python 03-subgrouping_samples.py 
+#	--final_table results/EPN_all_data.tsv 
+#	--subgroup_table results/EPN_all_data_withsubgroup.tsv 
+#	--temp_folder_name subgrouping_samples   ## This folder will keep all the histograms to view  the outliers but can  be deleted  after review
+
+
+## Importing modules to be used
+import pandas as pd
+import numpy as np 
+import sys
+import argparse
+from matplotlib import pyplot as plt
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--final_table', required = True,
+                    help = 'path to the notebook with all EPN columns')
+parser.add_argument('--subgroup_table', required = True,
+		    help = 'output table  with subgroup')
+parser.add_argument('--temp_folder_name', required = True,
+                    help = 'outout table  with subgroup')
+args = parser.parse_args()
+
+# REading in input table  
+EPN_final = pd.read_csv(args.final_table, sep="\t")
+# Adding a column with empty strings so subgroup name can be added
+EPN_final["subgroup"] = ""
+
+# File to write all the data to
+outfile = open(args.subgroup_table, "w")
+
+
+# This function takes in every row from EPN_final, 
+#	subgroup name to be returned(if values are higher than threshold),
+#       column values from EPN_final table  that need to be assessed for sub-grouping
+#If values higher than threshold and last column empty, this will return the subgroupname
+#	If there are other subgroup names already, it  appends the current subgroup name and returns
+def subgroup_func(row, subgroupname, column_values):
+    current_subgroup = row["subgroup"]
+    for columnname, value in column_values:
+        if row[columnname] > value:
+            if current_subgroup == '':
+                current_subgroup = subgroupname
+            elif subgroupname in current_subgroup.split(","):
+                pass
+            else:
+                current_subgroup = current_subgroup + "," + subgroupname
+
+    return(current_subgroup)
+
+## There is no set thresholds for gene expression outliers, so creating histograms for each of them.  
+## If gene expression Z-score is very high(outliers), that gene might be overexpressed in that sample and
+## helps the sample to be subgrouped accordingly??
```

Comment:
This seems like an important question. I am not sure I am the best one to address it, but is this generally a criterion used for subtyping EPNs? We might want to note in the output when we are typing based on specific features like fusions and losses, rather than by a continuous figure like expression. I would love to hear other thoughts on this.

Reply:
I had this question too.. the ticket says `genes over-expressed relative to other samples`. I created a histogram file with all the Z-scores and manually inspected and where ever I saw clear outliers, I tried to set a threshold just before that. I am open to ideas if there is a better way to do this.

Also, I noticed that histograms were empty from yesterday's commit, so I updated those. I will leave it up to you guys if you want to keep them in the EPN folder or write to a scratch folder.

Reply:
> I had this question too.. the ticket says `genes over-expressed relative to other samples`. I created a histogram file with all the Z-scores and manually inspected and where ever I saw clear outliers, I tried to set a threshold just before that. I am open to ideas if there is a better way to do this.

In other modules, we set the threshold as z-scored expression value > 3 for overexpression across the board.  Here's an example from `molecular-subtyping-embryonal`: https://alexslemonade.github.io/OpenPBTA-analysis/analyses/molecular-subtyping-embryonal/04-table-prep.nb.html#cns_nb-foxr2

In the non-ATRT/non-MB embryonal cases, it was generally pretty clear a gene could be called overexpressed. I would prefer we're conservative - use a uniform threshold and if we are unable to assign a molecular subtype based on that threshold, then we don't assign a molecular subtype.

Reply:
In this case, we probably will not need the histograms. I will work on removing those along with the other changes that Josh mentioned.

Reply:
Sounds good, thank you @tkoganti !